File: src/jvmTest/java/org/snakeyaml/engine/usecases/external_test_suite/EventRepresentation.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.usecases.external_test_suite;

import com.google.common.base.Splitter;
import org.snakeyaml.engine.v2.common.FlowStyle;
import org.snakeyaml.engine.v2.events.*;
import org.snakeyaml.engine.v2.nodes.Tag;

import java.util.List;

/**
 * Event representation for the external test suite
 */
public class EventRepresentation {

  private final Event event;

  public EventRepresentation(Event event) {
    this.event = event;
  }

  public String getRepresentation() {
    return event.toString();
  }


  public boolean isSameAs(String data) {
    List<String> split = Splitter.on(' ').splitToList(data);
    if (!event.toString().startsWith(split.get(0))) {
      return false;
    }
    /*
     * if (event instanceof DocumentStartEvent) { DocumentStartEvent e = (DocumentStartEvent) event;
     * if (e.isExplicit()) { if (split.size() != 2 || !split.get(1).equals("---")) return false; }
     * else { if (split.size() != 1) return false; } } if (event instanceof DocumentEndEvent) {
     * DocumentEndEvent e = (DocumentEndEvent) event; if (e.isExplicit()) { if (split.size() != 2 ||
     * !split.get(1).equals("...")) return false; } else { if (split.size() != 1) return false; } }
     */
    if (event instanceof MappingStartEvent) {
      CollectionStartEvent e = (CollectionStartEvent) event;
      boolean tagIsPresent = e.getTag() != null;
      String mapTag = Tag.MAP.getValue();
      if (tagIsPresent && !mapTag.equals(e.getTag())) {
        String last = split.get(split.size() - 1);
        if (!last.equals("<" + e.getTag() + ">")) {
          return false;
        }
      }
    }
    if (event instanceof SequenceStartEvent) {
      SequenceStartEvent e = (SequenceStartEvent) event;
      if (e.getTag() != null && !Tag.SEQ.getValue().equals(e.getTag())) {
        String last = split.get(split.size() - 1);
        if (!last.equals("<" + e.getTag() + ">")) {
          return false;
        }
      }
    }
    if (event instanceof NodeEvent) {
      NodeEvent e = (NodeEvent) event;
      if (e.getAnchor() != null) {
        int indexOfAlias = 1;
        if (event.getEventId().equals(Event.ID.SequenceStart)
            || event.getEventId().equals(Event.ID.MappingStart)) {
          CollectionStartEvent start = (CollectionStartEvent) event;
          if (start.getFlowStyle() == FlowStyle.FLOW) {
            indexOfAlias = 2;
          }
        }
        if (event instanceof AliasEvent) {
          if (!split.get(indexOfAlias).startsWith("*")) {
            return false;
          }
        } else {
          if (!split.get(indexOfAlias).startsWith("&")) {
            return false;
          }
        }
      }
    }
    if (event instanceof ScalarEvent) {
      ScalarEvent e = (ScalarEvent) event;
      if (e.getTag() != null) {
        String tag = e.getTag();
        ImplicitTuple implicit = e.getImplicit();
        if (implicit.bothFalse()) {
          if (!data.contains("<" + e.getTag() + ">")) {
            return false;
          }
        }
      }
      String end = e.getScalarStyle() + e.escapedValue();
      return data.endsWith(end);
    }
    return true;
  }
}

File: src/commonTest/kotlin/org/snakeyaml/engine/test_suite/EventRepresentation.kt
package org.snakeyaml.engine.test_suite

import org.snakeyaml.engine.v2.common.FlowStyle
import org.snakeyaml.engine.v2.events.*
import org.snakeyaml.engine.v2.nodes.Tag

/**
 * Event representation for the external test suite
 */
class EventRepresentation(private val event: Event) {
    val representation: String
        get() = event.toString()

    fun isSameAs(data: String): Boolean {
        val split: List<String> =data.split(" ")
        if (!event.toString().startsWith(split[0])) {
            return false
        }

//        when (event) {
//            is CollectionEndEvent -> TODO()
//            is CommentEvent -> TODO()
//            is DocumentEndEvent -> TODO()
//            is DocumentStartEvent -> TODO()
//            is NodeEvent -> TODO()
//            is StreamEndEvent -> TODO()
//            is StreamStartEvent -> TODO()
//        }

        if (event is MappingStartEvent) {
            val e = event as CollectionStartEvent
            val tagIsPresent = e.tag != null
            val mapTag = Tag.MAP.value
            if (tagIsPresent && mapTag != e.tag) {
                val last = split[split.size - 1]
                if (last != "<${e.tag}>") {
                    return false
                }
            }
        }
        if (event is SequenceStartEvent) {
            val e = event
            if (e.tag != null && Tag.SEQ.value != e.tag) {
                val last = split[split.size - 1]
                if (last != "<${e.tag}>") {
                    return false
                }
            }
        }
        if (event is NodeEvent) {
            if (event.anchor != null) {
                var indexOfAlias = 1
                if (event.eventId == Event.ID.SequenceStart || event.eventId == Event.ID.MappingStart) {
                    val start = event as CollectionStartEvent
                    if (start.flowStyle === FlowStyle.FLOW) {
                        indexOfAlias = 2
                    }
                }
                if (event is AliasEvent) {
                    if (!split[indexOfAlias].startsWith("*")) {
                        return false
                    }
                } else {
                    if (!split[indexOfAlias].startsWith("&")) {
                        return false
                    }
                }
            }
        }
        if (event is ScalarEvent) {
             if (event.tag != null) {
                val implicit = event.implicit
                if (implicit.bothFalse()) {
                    if (!data.contains("<${event.tag}>")) {
                        return false
                    }
                }
            }
            val end = event.scalarStyle.toString() + event.escapedValue()
            return data.endsWith(end)
        }
        return true
    }
}
